### ISSUE TITLE
For high DPI device setting ZoomLevel doesn't work immediately in WPF.

### ISSUE THREAD

**Issue by mookrs** [cefsharp/CefSharp#738]
Hi.

I'm developing on my rMBP which has 2560×1600 resolution, and our application will run on a 1920x1200 resolution tablet. As every element is too small by default, I tried setting `ZoomLevel="4"`, but the page didn't render as I supposed. This is how it looks:

![](http://i.imgur.com/kzJsEQs.jpg)

However, after I clicked `Zoom In` menu, it rendered like what I wish:

![](http://i.imgur.com/J7qlwSS.jpg)

I see #619 and find it works well, maybe I will use this solution.

But I don't know why the `ZoomLevel` doesn't work, is it a bug or maybe I did something wrong? In addition, will there be a Property like `AutoZoom` in the future? (Sorry I'm not familiar with GitHub and English.)

My environment is CefSharp.Wpf 37.0.0, Windows 8.1. I changed the scale of screen to 200 percent.

Thanks very very much. :)

**Comment by jornh** [comment]
> But I don't know why the ZoomLevel doesn't work, is it a bug or maybe I did something wrong?

Quite possibly a bug. High DPI has not been tested that well as none of the core CefSharp developers have a High DPI environment. Apart from #619 yours is the only input we have gotten so far on how High DPI works/doesn't work since CefSharp based on CEF 3.* went out of beta.

If #619 works improves matters for you as well that counts a validation of it too :+1: Thanks :smile: 

@bjarteskogoy (@amaitland or others) any (new) insights to add - or should we simply merge #619 based on the above?

**Comment by mookrs** [comment]
@jornh Thank you~ It's really nice of you~ :smile:

If your team needs more test or details, I would be glad to help.

**Comment by amaitland** [comment]
> @bjarteskogoy (@amaitland or others) any (new) insights to add - or should we simply merge #619 based on the above?

@jornh As the codebase is getting larger, I'd prefer to look at implementing this sort of behavior as an `Attached Property`, that way those requiring it simply add some xaml. Just need to expose a method that provides the zoom level based on dpi.

> If your team needs more test or details, I would be glad to help.

@mookrs Appreciate the offer! Any chance you could take the `39.0.0-pre01` nuget package for a spin and add high dpi support to your app manifest? Background info here https://github.com/cefsharp/CefSharp/pull/664#issuecomment-65660436.

I believe `39` has build in support for high dpi, so the hope is that adding an app manifest will do the trick.

**Comment by amaitland** [comment]
Other possible options include adding `--high-dpi-support=1` command line option

Background: https://code.google.com/p/chromiumembedded/issues/detail?id=1359#c6

**Comment by amaitland** [comment]
https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/ManagedCefBrowserAdapter.h#L463

For reference I believe setting `ZoomLevel="4"` in xaml would result in the initial call to `SetZoomLevel` being too early, and `cefHost` would still be a `nullptr`.

**Comment by amaitland** [comment]
If your using `.Net 4.5` then it's probably worth trying the `Delay` binding property.

http://www.jonathanantoine.com/2011/09/21/wpf-4-5-part-4-the-new-bindings-delay-property/

**Comment by mookrs** [comment]
I used `CefSharp.MinimalExample` solution for testing. First, I update nuget package to `CefSharp.Wpf 39.0.0-pre01` `CefSharp.OffScreen 39.0.0-pre01` `CefSharp.WinForms 39.0.0-pre01`. Second, I added a `app.manifest` file to the `CefSharp.MinimalExample.Wpf` project. I modified and added these lines into the default `app.manifest` (It was as same as jimnoble's commit 3885aec88b4007f8a3ff44db8dbead4a461ef956):

```xml
<asmv1:assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1" 
                xmlns:asmv1="urn:schemas-microsoft-com:asm.v1" 
                xmlns:asmv2="urn:schemas-microsoft-com:asm.v2"
                xmlns:asmv3="urn:schemas-microsoft-com:asm.v3"  
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
```

````xml
<asmv3:application>
  <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">
    <dpiAware>True/PM</dpiAware>
  </asmv3:windowsSettings>
</asmv3:application>
```

The debug target was set to `x64`. I rebuilted the solution, but nothing changed:

![](http://i.imgur.com/uNzmMhZ.jpg)

I also set the `CefSharp.MinimalExample.WinForms` as start project without a `app.manifest`, it looks not clear, but the size is fit:

![](http://i.imgur.com/aa4qf4v.jpg)

I will do more test later. If I have made any mistake, please point out. Thanks a lot!

**Comment by amaitland** [comment]
@mookrs Appreciate you taking the time to test!

It sounds like you've done everything correctly :+1: You can check this branch for reference https://github.com/cefsharp/CefSharp.MinimalExample/tree/test/dpi-manifest

We may actually have to remove our custom scaling.

**Comment by amaitland** [comment]
It's probably worth asking on the `CEF` forum what's required to get `OSR rendering` to work with high dpi.

http://www.magpcss.org/ceforum/viewtopic.php?f=6&t=12366

**Comment by mookrs** [comment]
@amaitland I used the new tree for testing, the browser rendered as before. Elements were still small.

As far as I know, WPF is resolution independence. I think maybe it doesn't have to use `app.manifest` file for DPI purpose. 

By the way, yesterday I tested `CefSharp.Wpf.Example` in `CefSharp-master`, after I clicked a `target="_blank"` link, it appeared a new WPF window with the new page. The page rendered well. 

Em..This is really a confusing problem... :joy:

**Comment by amaitland** [comment]
> As far as I know, WPF is resolution independence. I think maybe it doesn't have to use app.manifest file for DPI purpose.

As the `WPF` control goes it's actually a little more complex than your typical `WPF` control. It's not a native control per say, it's actually an image with virtual key presses and mouse events that get sent to the underlying `CEF` browser, which are then translated on screen and rendered as a new image. So the task at hand is getting `CEF` to be dpi aware.

> update 2 --high-dpi-support=1 worked in neither WinForm nor WPF.

Check the latest commit to https://github.com/cefsharp/CefSharp.MinimalExample/tree/test/dpi-manifest for command line options

**Comment by amaitland** [comment]
(There's no notification when a comment is edited, so best to post follow up comments)

**Comment by mookrs** [comment]
@amaitland Got it. I will try the new project. Thanks!

**Comment by mookrs** [comment]
@amaitland I tried serveral times and changed paramer values, none of them worked. (such as use `settings.CefCommandLineArgs.Add("high-dpi-support", "1/0");` uncomment `settings.CefCommandLineArgs.Add("force-device-scale-factor", "1/0");` test `https://www.google.com`)

My `chrome://version` info is:

```
CEF	3.2171.1949
Chromium	39.0.2171.65
OS	Windows
WebKit	537.36
JavaScript	3.29.88.17
Flash	
User Agent	Mozilla/5.0 (Windows NT 6.2; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.65 Safari/537.36
Command Line	"C:\Users\<My Username>\Desktop\CefSharp.MinimalExample-test-dpi-manifest\CefSharp.MinimalExample.Wpf\bin\x64\Debug\CefSharp.MinimalExample.Wpf.exe" --browser-subprocess-path=CefSharp.BrowserSubprocess.exe --no-sandbox --lang=en-US --high-dpi-support=1
Module Path	C:\Users\<My Username>\Desktop\CefSharp.MinimalExample-test-dpi-manifest\CefSharp.MinimalExample.Wpf\bin\x64\Debug\libcef.dll
Cache Path	C:\Users\<My Username>\AppData\Local\Temp\scoped_dir9012_472
```

Thank you! :+1:

**Comment by amaitland** [comment]
@mookrs Appreciate you taking the time :+1: 

I'm out of ideas at the moment, if I think of anything I'll report back.

**Comment by mookrs** [comment]
@amaitland I'm appreciate your help, waiting for your good news!

**Comment by jornh** [comment]
@mookrs just one small thought/question: Have you tried comparing with `cefclient.exe` from the **Test App** download from http://cefbuilds.com ? 

If you run it with `cefclient --multi-threaded-message-loop --off-screen-rendering-enabled` as command line parameters I think that would get as close to how it's running in `CefSharp.Wpf` and we get it a bit more isolated to if the issue is with upstream CEF or CefSharp's Wpf implementation.

**Comment by mookrs** [comment]
@jornh I just made a test, run the command in a CMD window. The results for `Windows 32bit Chromium 41.0.2271.0 gd3cf92ca Test App` and `Windows 32bit Chromium 39.0.2171.95 Test App` were like `CefSharp.MinimalExample.WinForms`, fit size but not clear.

I tried disable high DPI scaling in `cefclient.exe`'s property options, this made it behave like WinForm application added the `app.manifest` file, which at first time web page was overflow of the window, after resizing it bacame normal and rendered very sharply.

**Comment by mookrs** [comment]
@amaitland I find this link they say  [Windows --high-dpi-support flag is going away - Now use the registry.](https://groups.google.com/a/chromium.org/forum/#!searchin/chromium-dev/dpi/chromium-dev/U_K1NHuMLic/L3zWyliK3woJ),  it seems that the switch was removed in branch 1750 or later.

Hope this could be useful!

**Comment by jornh** [comment]
> I tried disable high DPI scaling in cefclient.exe's property options, this made it behave like WinForm application added the app.manifest file, which at first time web page was overflow of the window, after resizing it bacame normal and rendered very sharply.

So in other words AND just to be 100% sure (yes I'm confused because I don't have the issue in front of me :stuck_out_tongue_winking_eye:); this is when running `cefclient.exe` with `--off-screen-rendering-enabled` and then it behaves with issues until first resize ... or ... ? Because then I think we need to handle this as an upstream CEF project issue, so you need to go there, see if it's already reported in the issue tracker, possibly ask for advice in the CEF forum and file an issue if required.

**Comment by jornh** [comment]
Ah well I guess (some of it) is this already known CEF issue https://code.google.com/p/chromiumembedded/issues/detail?id=1445&q=dpi referred to in #619 ... but as I said I'm confused :smiling_imp: - marking as `upstream` I don't know if there's anything we can/should do currently in the CefSharp codebase

**Comment by mookrs** [comment]
@jornh From your link I saw the issue, it was as same as mine, thank you for your reminding. :grinning:

Actually I just want my WPF application to work well... Because of the unknown problem, maybe I have to wait for some days. HiDPI is always a terrible problem...

**Comment by amaitland** [comment]
> tried disable high DPI scaling in cefclient.exe's property options

Passing in command line args? Or some other way?

**Comment by mookrs** [comment]
@amaitland I'm sorry I didn't express it clearly. In fact, I right-clicked the application’s exe and then click Properties. On the Compatibility tab, select Disable Display Scaling On High DPI Settings, and then click OK.

**Comment by amaitland** [comment]
@mookrs Thanks :+1:

**Comment by mookrs** [comment]
@amaitland You guys are really nice and diligent. Good luck! :smile:

**Comment by amaitland** [comment]
Background info for future reference http://magpcss.org/ceforum/viewtopic.php?f=6&t=12080&hilit=OnLoadingStateChange

**Comment by amaitland** [comment]
> @amaitland You guys are really nice and diligent. Good luck! :smile:

Thanks @mookrs! Really appreciate you taking the time to investigate all the possible options!

My plan is to implement `AutoZoom` as an attached property (codebase is getting rather large, so trying to split out pieces into smaller behaviors). Feel free to check it out if interested. https://github.com/amaitland/CefSharp/commit/a83e4804f5daabf41fdc7270947482a52a728dfc

**Comment by mookrs** [comment]
@amaitland Thanks for your work! Sorry for replying late during weekend. Today I checkouted this commit https://github.com/amaitland/CefSharp/commit/a83e4804f5daabf41fdc7270947482a52a728dfc, compiled and ran it. However, it behaved like `ZoomLevel`. The web page was small, unless clicked `Zoom In` Menu once. I'm afraid `AutoZoom` attached property maybe was also called too early. Would you mind checking where is wrong once more? As our project is not in a hurry, just take your time to fix it. Thanks again!

**Comment by amaitland** [comment]
@mookrs Thanks for taking it for a spin :+1: 

Perhaps the binding `Delay` is an option? Give the UI a chance to render.

**Comment by mookrs** [comment]
@amaitland I tried using `Delay` property in my way. I added a property named `IsAutoZoom` into BrowserTabViewModel, in its constructor I set `IsAutoZoom = true;`. In `BrowserTabView.xaml` I set `attachedProperties:ZoomBehaviour.AutoZoom="{Binding IsAutoZoom, Delay=1000}"`, but this didn't work...

**Comment by amaitland** [comment]
@mookrs If you just set the `ZoomLevel` property with a delay is that any different?

**Comment by mookrs** [comment]
@amaitland Nothing has changed, it bahaves like without `Delay`. :pensive:

**Comment by amaitland** [comment]
@mookrs If I change the AutoZoom attached property to hard coded at `ZoomLevel = 4` I get the result I'm expecting. Perhaps you could just set the `ZoomLevel` to the desired value in `FrameLoadStart`?

https://github.com/amaitland/CefSharp/commit/7c4ef7d9f23b54d93e3ba0b97fe85df4af7f31ba

**Comment by mookrs** [comment]
@amaitland Maybe I find something different. When I came into the application, the Google site rendered well! But... After I typed another site into address bar, the page rendered as before.

The previous code use `CreateNewTab(CefExample.DefaultUrl, true);`, the `ZoomLevel ` or `AutoZoom ` property maybe have affected for the `DefaultUrl`. After typing in another url, effect of the property disappears. Perhaps `GoCommand` or something else are the reason.

**Comment by mookrs** [comment]
@amaitland Base on https://github.com/amaitland/CefSharp/commit/7c4ef7d9f23b54d93e3ba0b97fe85df4af7f31ba , I just need to change here: https://github.com/bjarteskogoy/CefSharp/commit/d08df3d96a5f3402894cdbe38950c9663a5912b1 , then everything is fine.

**Comment by jornh** [comment]
In #619 Bjarte noted another related CEF issue:

> Good. Anyway, it looks like there is a bug in cef making it impossible to be sure you have the right zoom on the first page load: https://code.google.com/p/chromiumembedded/issues/detail?id=1467&sort=-id

**Comment by bjarteskogoy** [comment]
For WPF: https://github.com/cefsharp/CefSharp/pull/842

**Comment by mookrs** [comment]
@amaitland @bjarteskogoy Hi everyone, I come here again.

 I just build `v39.0.0` and set `ZoomLevel="4"`, but the `CefSharp.Wpf.Example` project still rendered like before —— It was very small at first time, after I clicked `Zoom In` menu, it rendered well.

So, what may I miss? Has `v39.0.0` solved the HiDPI problem? Thank you! :smile:

**Comment by amaitland** [comment]
> Has v39.0.0 solved the HiDPI problem?

No, #858 still needs work.

**Comment by mookrs** [comment]
@amaitland Got it. Waiting for your good news!

**Comment by amaitland** [comment]
@mookrs As a workaround, what happens if you call `SetZoomLevel(4.0F)` in `FrameLoadEnd`? I think part of the problem was the `DependencyProperty`. So if you call the method directly it works for the most part.

**Comment by mookrs** [comment]
@amaitland I used `browser.SetZoomLevel(4.0F);` in `FrameLoadEnd` event handler, before the page loaded end it was small, after the page loaded end it increased to the right zoom level. 

The problem is that I have to wait for the page loading end, I can see the page suddenly becomes bigger. And when I go to another website, it renders from small to big again.

**Comment by amaitland** [comment]
You may actually be able to set it in `FrameLoadStart`. (I've only quickly tested with the `2272` branch and it appears to be working, might be different for version `39.0.0` which is the `2171` branch).

```c#
private static void WebBrowserFrameLoadStart(object sender, FrameLoadStartEventArgs e)
{
	if (e.IsMainFrame)
	{
		var browser = (ChromiumWebBrowser)sender;
		browser.SetZoomLevel(4.0f);
	}
}
```

**Comment by mookrs** [comment]
@amaitland Wow, what you reminded me was actually I just have done! I've tried it, worked very well in `39.0.0`! :+1:

**Comment by mookrs** [comment]
@amaitland I used
```
<cefSharp:ChromiumWebBrowser x:Name="browser" FrameLoadStart="Browser_OnFrameLoadStart" ... />
``` 
and
 ```
private void Browser_OnFrameLoadStart(object sender, FrameLoadStartEventArgs e)
{
    browser.SetZoomLevel(4.0F);
}
```
Worked fine. Really appreciate your work!

**Comment by amaitland** [comment]
Great, glad there's a workaround :+1:

**Comment by canton7** [comment]
It's worth noting that, for the WPF control, you *must* call `SetZoomLevel` from the `FrameLoadStart` event, rather than using the `ZoomLevel` dependency property. The dependency property requires that you only change it from the UI thread, and `FrameLoadStart` is invoked on a background thread. If you dispatch to the UI thread here to set `ZoomLevel`, you get all sorts of threading issues (random render problems, no Javascript, etc).

This means you have to use `ChromiumWebBrowser` and not `IWpfWebBrowser`. It also means that you'll likely not be able to use `ZoomLevel` at all, since now it will be out of date, and it will only call its 'changed' handler when its value actually changes (so calling `SetZoomLevel(1.0)`, and `ZoomLevel = 0`, will set to 1.0 but not set back to 0).

**Comment by amaitland** [comment]
The complexity of the

**Comment by amaitland** [comment]
I'm thinking of removing `ZoomLevel` and adding both `GetZoomLevel` and `SetZoomLevel` to the `IWebBrowser` interface. This would need to be done on the `2272` branch as it's a breaking change.

Thoughts?

**Comment by canton7** [comment]
`ZoomLevel` would be useful I think - it's bindable, so means it fits nicely into the MVVM pattern. I only had to abandon it here because of the workaround required for this issue. In an ideal world, I'd just use the `ZoomLevel` property.

**Comment by amaitland** [comment]
As you've discovered there are threading issues. It's fine if you bind a command and let the user execute. Soon as you interact with it from an eventhandler it becomes problematic, it also doesn't reflect the underlying `CEF API` particularly well. The `WinForms` implementation for example wraps two methods in a property so it's not ideal.

`ZoomLevel` was my first contribution to the `CefSharp` project, and as such doesn't take the `CEF threading model` into account.

**Comment by canton7** [comment]
If the user wants to randomly change the Zoom Level, what thread should they do it from?

**Comment by mookrs** [comment]
By the way, my workaround has a littile problem, it will invoke everytime when `FrameLoadStart`. Hope this can improve.

```csharp
public MainWindow()
{
    InitializeComponent();
    browser.FrameLoadStart += Browser_FrameLoadStart;
}

private void Browser_FrameLoadStart(object sender, FrameLoadStartEventArgs e)
{
    browser.SetZoomLevel(4.0);
    browser.FrameLoadStart -= Browser_FrameLoadStart;
}
```

**Comment by bjarteskogoy** [comment]
Setting the ZoomLevel is only a temporary solution. It believe also does not scale the scrollbars, which make them quite unusable on touch devices with 150%+ scaling.

We need to get this one fixed https://github.com/cefsharp/CefSharp/pull/937.

I believe it should be enough to implement IRenderWebBrowser.GetScreenInfo() and provide the correct scaling, combined with the writablebitmap when dpi is > 100% (due to interopbitmap having a fixed dpi).

**Comment by amaitland** [comment]
> Setting the ZoomLevel is only a temporary solution.

For clarity perhaps that should be `Setting ZoomLevel in FrameLoadStart` is a temporary solution?

**Comment by amaitland** [comment]
> If the user wants to randomly change the Zoom Level, what thread should they do it from?

`ZoomLevel` is a dependency property, so standard rules apply, you have to use the `Dispatcher` to change it. And my guess is that causes some of the problems.

See http://magpcss.org/ceforum/apidocs3/projects/%28default%29/CefBrowserHost.html#SetZoomLevel%28double%29 

Remember the `CEF UI` thread is different to the `WPF UI` thread. So the call currently will eventually marshal onto the correct thread. Perhaps explicitly executing using `CefPostTask` will resolve some of the issues.

**Comment by canton7** [comment]
My question was geared towards trying to find out what thread `SetZoomLevel` *should* be called on - as my  comment [here](https://github.com/cefsharp/CefSharp/issues/738#issuecomment-94787470) mentions, calling it on the UI thread from the `FrameLoadStart` event causes all sorts of weird issues, whereas calling it synchronously in this event (from the thread on which the event is fired) is fine. 

It sounds to me as if `OnZoomLevelChanged` (which is called when `ZoomLevel` is set) should dispatch its call to `SetZoomLevel` to the `CEF UI` thread.

**Comment by amaitland** [comment]
If you'd read the `CEF API` link I posted above you'd have seen `If called on the UI thread the change will be applied immediately. Otherwise, the change will be applied asynchronously on the UI thread.`

**Comment by canton7** [comment]
I saw that, but the evidence I'm seeing implies that this isn't the case here. If I:

1. Register a `FrameLoadStart` (or indeed a `FrameLoadEnd`) event handler
2. Inside this handler, dispatch to the (WPF) UI thread, and call `ZoomLevel = whatever`

then CEF does some really random stuff, including (but not limited to):

1. Rending images as massive
2. Not rendering images at all
3. Ignoring some, but not all, margin/padding

What it actually decides to do varies from attempt to attempt. I (perhaps incorrectly) interpreted this as a threading issue (`SetZoomLevel being called on the wrong thread), and starting calling `SetZoomLevel` synchronously from inside the event handler instead.

**Comment by amaitland** [comment]
Once #961 is merged, it'll be much easier to marshall calls onto different `CEF` threads, so you can always experiment running it explicitly on the `UI` thread to see if it makes a difference.

**Comment by amaitland** [comment]
`High DPI` support in `WPF` has been improved. If there's still a problem then please open a new issue.

**Comment by kevinvan44** [comment]
@mookrs， appreciate your work! But I have a question and wanna ask you. Why did you set the zoomLevel to 4? I am currently using your method, however, I found that 3.8 might be better, since the content pulled from my web server is fixed-size. Moreover, I tried different DPI. It looks like there is not a fixed relation between the zoomLevel and the DPI scale.

**Comment by kevinvan44** [comment]
And one more question. After the contents shown, I changed the DPI of the computer. Guess what, the whole contents are empty. I have to restart my application. That is weired...

**Comment by mookrs** [comment]
@kevinvan44 For your first question, I guess it is because my computer having 2560×1600 resolution, which is about 4 times than normal (1280×800). I think 3.8 is fine according to your situation.

For your other questions, I'm afraid I can't give you more advises, since I have quitted WPF relevant
development for a long time, I even don't know the present situation of our project. 😅

Good luck to you!
